ArduCopter/GCS_Mavlink.cpp

```diff
@@ -1194,6 +1194,34 @@ void GCS_MAVLINK_Copter::handleMessage(mavlink_message_t* msg)
             break;
 #endif
 
+        case MAV_CMD_DO_WINCH:
+            // param1 : winch number (ignored)
+            // param2 : action (0=release, 1=length control, 2=rate control). See WINCH_ACTIONS enum.
+            if(!copter.g2.winch.enabled()) {
+                result = MAV_RESULT_FAILED;
+            } else {
+                result = MAV_RESULT_ACCEPTED;
+                switch ((uint8_t)packet.param2) {
+                    case WINCH_DISABLE:
+                        copter.g2.winch.relax();
+                        break;
+                    case WINCH_LENGTH_CONTROL: {
+                        float desired_length = copter.g2.winch.get_line_length() + packet.param3;
+                        copter.g2.winch.release_length(desired_length);
+                        break;
+                    }
+                    case WINCH_RATE_CONTROL: {
+                        float desired_rate = packet.param4;
+                        copter.g2.winch.set_desired_rate(desired_rate);
```

Comment:
If a rate is requested that is outside the possible rate of the winch it should be a rejected command instead of simply clamping at the max rate. The problem with accepting it, is that it might convince a operator that more or less length has been deployed/retracted then actually has, and could cause the operator to plan a path/command a travel that would actually snag the line.